Update devices each 1 second until chain is not created

AlexeyBarabash opened the pull request:
Fixes https://github.com/brave/brave-browser/issues/2742 , https://github.com/brave/brave-browser/issues/2782 .

Improves response time on sync creation. 

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:
1. On Device1 go to chrome://sync/
2. Press "Start new chain" => "Computer" => "Copy code"
3. On device 2 go to chrome://sync/
4. "Enter sync chain code" button
5. Paste the code
6. "Confirm sync code" button
7. Device1 gets created a sync chain not so late (<5 sec) after Device2

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

darkdh commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -480,6 +483,7 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
           actually_merged);
       contains_only_one_device = sync_devices->size() < 2 &&
         record->action == jslib::SyncRecord::Action::A_DELETE;
+      have_two_or_more_after_merge = actually_merged && sync_devices->size() >= 2;
```
we can put `s/had_less_than_two/waiting_for_second_device` here and no need to check > 2 case because StartLoop should only be initialized once.
Should look like this
`second_device_connected = waiting_for_second_device  && actually_merged && sync_devices->size() == 2`

AlexeyBarabash replied:
thanks, simplified this.

darkdh commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -489,6 +493,10 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
   if (contains_only_one_device)
     OnResetSync();
+
+  if (had_less_than_two && have_two_or_more_after_merge) {
```
per https://github.com/brave/brave-core/pull/1188#pullrequestreview-188147327
we should `StopLoop` first 
And should be sufficient checking one condition, see the comment above

AlexeyBarabash replied:
I think, we cannot skip send `StopLoop` because `base::RepeatingTimer` has a logic to set a new interval for existing timer:
https://cs.chromium.org/chromium/src/base/timer/timer.cc?type=cs&g=0&l=120
https://cs.chromium.org/chromium/src/base/timer/timer.cc?type=cs&g=0&l=137

darkdh commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -284,7 +284,8 @@ void BraveSyncServiceImpl::BackgroundSyncStarted(bool startup) {
     bookmark_change_processor_->Start();
 
-  StartLoop();
+  const bool initial_updates = sync_prefs_->GetSyncDevices()->size() <= 1;
```
s/initial_updates/waiting_for_second_device/ would be easier to understand why we are doing this

AlexeyBarabash replied:
thanks, fixed.

darkdh commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -603,10 +611,13 @@ void BraveSyncServiceImpl::SendDeviceSyncRecord(
   timer_->Start(FROM_HERE,
-                  base::TimeDelta::FromSeconds(kCheckUpdatesIntervalSec),
+                  base::TimeDelta::FromSeconds(
+                    initial_frequent_updates ? kCheckInitialUpdatesIntervalSec :
```
s/initial_updates/waiting_for_second_device/ because that is the only case we fetch every 1 second

AlexeyBarabash replied:
thanks, fixed.

AlexeyBarabash commented:
Pushed the fix. Will squash and rebase after approval.

darkdh commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -284,7 +284,8 @@ void BraveSyncServiceImpl::BackgroundSyncStarted(bool startup) {
     bookmark_change_processor_->Start();
 
-  StartLoop();
+  const bool waiting_for_second_device = sync_prefs_->GetSyncDevices()->size() <= 1;
```
lint: <= 80 chars

bridiver replied:
fyi - https://github.com/brave/brave-browser/pull/2743

darkdh commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -603,10 +610,13 @@ void BraveSyncServiceImpl::SendDeviceSyncRecord(
   timer_->Start(FROM_HERE,
-                  base::TimeDelta::FromSeconds(kCheckUpdatesIntervalSec),
+                  base::TimeDelta::FromSeconds(
+                    waiting_for_second_device ? kCheckInitialUpdatesIntervalSec :
```
lint: <= 80 chars

bridiver commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -463,6 +464,7 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
   bool contains_only_one_device = false;
 
   auto sync_devices = sync_prefs_->GetSyncDevices();
+  bool waiting_for_second_device = sync_devices->size() < 2;
```
can you add a method for this? There are two checks that both check the same condition, but with different values `<= 1` vs `< 2` and it's confusing to read like that

bridiver commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -489,6 +491,11 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
   if (contains_only_one_device)
     OnResetSync();
+
+  if (waiting_for_second_device && sync_devices->size() >= 2) {
```
how can this ever be true? `sync_devices->size() < 2 && sync_devices->size() >= 2` ?

darkdh replied:
before applying resolved records, device size was 1 and after that, it should be 2 or more depends on what resolved records we receive

bridiver replied:
I see, the code that was collapsed updates the sync_device count?

bridiver commented:
@AlexeyBarabash can you please add a test for this?

AlexeyBarabash commented:
@darkdh @bridiver 
Thanks for the review.
Added a test and fixed lint issues.
Tried to run `npm run lint` but then understand https://github.com/brave/brave-browser/pull/2743 is not merged yet. In any way looking forward to start use it.

AlexeyBarabash commented:
Pushed a fix for https://github.com/brave/brave-core/pull/1188#discussion_r244385148 comment .

AlexeyBarabash commented:
Rebased and squashed commits.

bridiver commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -484,7 +487,7 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
           record->action == jslib::SyncRecord::Action::A_DELETE &&
           actually_merged);
-      contains_only_one_device = sync_devices->size() < 2 &&
+      contains_only_one_device = sync_devices->have_less_than_two() &&
```
I'm confused about this, if there's only one device and it has been deleted, then there are no devices, not one device

bridiver commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -496,6 +499,11 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
   if (contains_only_one_device)
     OnResetSync();
+
+  if (waiting_for_second_device && !sync_devices->have_less_than_two()) {
```
I don't understand why we need this second check and the extra var. Why do we care what the previous size of the sync_devices was? It's also possible that both `this_device_deleted` and `contains_only_one_device` can both be true. The logic here just doesn't make sense to me

bridiver replied:
couldn't we just do something like this?
```
diff --git a/components/brave_sync/brave_sync_service_impl.cc b/components/brave_sync/brave_sync_service_impl.cc
index cb76609e..6fa11216 100644
--- a/components/brave_sync/brave_sync_service_impl.cc
+++ b/components/brave_sync/brave_sync_service_impl.cc
@@ -464,9 +464,8 @@ BraveSyncServiceImpl::PrepareResolvedPreferences(const RecordsList& records) {
 }
 
 void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
-  const std::string this_device_id = sync_prefs_->GetThisDeviceId();
   bool this_device_deleted = false;
-  bool contains_only_one_device = false;
+  bool this_device_added = false;
 
   auto sync_devices = sync_prefs_->GetSyncDevices();
   for (const auto &record : records) {
@@ -480,22 +479,23 @@ void BraveSyncServiceImpl::OnResolvedPreferences(const RecordsList& records) {
           record->syncTimestamp.ToJsTime()),
           record->action,
           &actually_merged);
-      this_device_deleted = this_device_deleted ||
-        (record->deviceId == this_device_id &&
-          record->action == jslib::SyncRecord::Action::A_DELETE &&
-          actually_merged);
-      contains_only_one_device = sync_devices->size() < 2 &&
-        record->action == jslib::SyncRecord::Action::A_DELETE &&
-          actually_merged;
+      if (record->deviceId == sync_prefs_->GetThisDeviceId() &&
+          actually_merged) {
+        this_device_deleted =
+            record->action == jslib::SyncRecord::Action::A_DELETE;
+        this_device_added =
+            record->action == jslib::SyncRecord::Action::A_CREATE;
+      }
     }
   } // for each device
 
   sync_prefs_->SetSyncDevices(*sync_devices);
 
-  if (this_device_deleted)
+  if (sync_devices->size() == 0)
     ResetSyncInternal();
-  if (contains_only_one_device)
-    OnResetSync();
+  else if (sync_devices->size() == 1) {
+    StartLoop(kCheckInitialUpdatesIntervalSec);
+  }
 }
 
 void BraveSyncServiceImpl::OnSyncPrefsChanged(const std::string& pref) {
```

bridiver commented on components/brave_sync/brave_sync_service_impl.cc:
```diff
@@ -610,10 +618,14 @@ void BraveSyncServiceImpl::SendDeviceSyncRecord(
+static const int64_t kCheckInitialUpdatesIntervalSec = 1;
 
-void BraveSyncServiceImpl::StartLoop() {
+void BraveSyncServiceImpl::StartLoop(const bool waiting_for_second_device) {
```
maybe `use_initial_update_interval`?

AlexeyBarabash commented:
Pushed fixes.

AlexeyBarabash commented:
Rebased to current master.

bridiver commented on components/brave_sync/sync_devices.h:
```diff
@@ -43,6 +43,7 @@ class SyncDevices {
    std::unique_ptr<base::Value> ToValueArrOnly() const;
    std::string ToJson() const;
    size_t size() const { return devices_.size(); }
+   bool have_less_than_two() const { return size() < 2; }
```
just a nit, but can we make this `has_second_device` and reverse the logic? I think the purpose would be more intuitive and `have_less_than_two` just sounds weird to me
